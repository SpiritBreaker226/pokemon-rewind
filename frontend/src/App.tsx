import React from 'react'

import { Grid, Typography } from '@material-ui/core'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'

import FetchData from './components/FetchData'
import CardTable from './components/CardTable'
import Search from './components/Search'
import Control from './components/Control'
import ErrorBoundary from './components/ErrorBoundary'

import { AppProvider } from './contexts/AppContext'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    appBody: {
      paddingBottom: theme.spacing(2),
      paddingLeft: theme.spacing(2),
      paddingRight: theme.spacing(2),
    },

    headerActions: {
      marginTop: theme.spacing(1),
    },

    header: {
      backgroundColor: theme.palette.secondary.main,
      padding: theme.spacing(2),
    },
  })
)

function App() {
  const classes = useStyles()

  return (
    <main>
      <ErrorBoundary>
        <AppProvider>
          <header className={classes.header}>
            <Grid container>
              <Grid item xs={4}>
                <Typography variant="h1">Pokemon Rewind</Typography>
              </Grid>
              <Grid item xs={4} className={classes.headerActions}>
                <Search />
              </Grid>
              <Grid item xs={4} className={classes.headerActions}>
                <Control />
              </Grid>
            </Grid>
          </header>

          <div className={classes.appBody}>
            <CardTable />
            <FetchData />
          </div>
        </AppProvider>
      </ErrorBoundary>
    </main>
  )
}

export default App
