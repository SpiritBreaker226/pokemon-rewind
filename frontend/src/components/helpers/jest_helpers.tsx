import React, { FunctionComponent } from 'react'
import UserEvent from '@testing-library/user-event'
import { within, waitForElementToBeRemoved } from '@testing-library/react'

import { Card } from '../../types/Cards'
import { Direction, Method } from '../../types/CardQuery'
import { Pagination } from '../../types/Pagination'
import { Action } from '../../types/Actions'
import { Search } from '../../types/Search'

import { AppContext } from '../../contexts/AppContext'

export const cards: Card[] = [
  {
    attributes: {
      name: 'Lightning Energy',
      supertype: 'Energy',
      subtype: 'Basic',
      number: '128',
      artist: 'Keiji Kinebuchi',
      rarity: 'Rare',
      series: 'Base',
      set_code: 'base4',
    },
  },
  {
    attributes: {
      name: 'Jigglypuff',
      national_pokedex_number: 39,
      supertype: 'Pokémon',
      subtype: 'Basic',
      hp: 60,
      converted_retreat_cost: 1,
      number: '77',
      artist: 'Kagemaru Himeno',
      rarity: 'Common',
      series: 'Base',
      set_code: 'base4',
    },
  },
  {
    attributes: {
      name: 'Squirtle',
      national_pokedex_number: 7,
      supertype: 'Pokémon',
      subtype: 'Basic',
      hp: 40,
      converted_retreat_cost: 1,
      number: '93',
      artist: 'Mitsuhiro Arita',
      rarity: 'Common',
      series: 'Base',
      set_code: 'base4',
    },
  },
]

export interface TestingState {
  cards?: Card[]
  pagination?: Pagination
  urlToEndpoint?: string
  search?: Search
  method?: Method
  errorMessage?: string
  isLoading?: boolean
  sorting?: {
    fieldName: string | undefined
    dir: Direction
  }
}

export interface MakeWrapperProps {
  state: TestingState
  dispatch?: (action: Action) => void
}

export const MakeWrapper: FunctionComponent<MakeWrapperProps> = ({
  state,
  dispatch,
  children,
}) => (
  <AppContext.Provider
    value={{
      state: {
        cards: [],
        pagination: {
          rows_per_page: 10,
          page_total: 0,
          current_page: '1',
        },
        urlToEndpoint: '',
        search: {
          toggle: false,
          value: {
            name: '',
            hp: '',
            rarity: '',
          },
          field: 'name',
        },
        method: 'GET',
        errorMessage: '',
        sorting: {
          fieldName: undefined,
          dir: undefined,
        },
        isLoading: false,
        ...state,
      },
      dispatch: dispatch || (() => null),
    }}
  >
    {children}
  </AppContext.Provider>
)

export const selectMaterialUiSelectOption = async (
  selectElement: HTMLElement,
  optionText: string
) =>
  new Promise((resolve) => {
    if (!selectElement || !selectElement.parentNode) {
      throw new Error(
        'Unable to find select element for selectMaterialUiSelectOption'
      )
      return
    }

    // The the button that opens the dropdown, which is a sibling of the input
    const selectButton = selectElement.parentNode.querySelector('[role=button]')

    if (!selectButton) {
      throw new Error(
        'Unable to find select button for selectMaterialUiSelectOption'
      )
      return
    }

    // Open the select dropdown
    UserEvent.click(selectButton)

    // Get the dropdown element. We don't use getByRole() because it includes <select>s too.
    const listbox = document.body.querySelector<HTMLElement>('ul[role=listbox]')

    if (!listbox) {
      throw new Error('Unable to find listbox for selectMaterialUiSelectOption')
      return
    }

    // Click the list item
    const listItem = within(listbox).getByText(optionText)
    UserEvent.click(listItem)

    // Wait for the listbox to be removed, so it isn't visible in subsequent calls
    waitForElementToBeRemoved(() =>
      document.body.querySelector('ul[role=listbox]')
    ).then(resolve)
  })
