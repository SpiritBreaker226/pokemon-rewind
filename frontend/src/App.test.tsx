import React from 'react'

import { waitFor, screen, render, fireEvent } from '@testing-library/react'

import axios from 'axios'

import { cards } from './components/helpers/jest_helpers'

import App from './App'

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

const pagination = {
  rows_per_page: 10,
  page_total: 20,
  current_page: '1',
}

describe('App', () => {
  describe('fetch card data', () => {
    beforeEach(async () => {
      mockedAxios.get.mockResolvedValue({
        data: { cards: { data: cards }, pagination },
      })

      render(<App />)

      await waitFor(() => screen.getByRole('heading'))
    })

    it('renders cards component', () => {
      expect(mockedAxios.get).toHaveBeenCalledTimes(1)
      expect(mockedAxios.get).toHaveBeenCalledWith(
        `${process.env.REACT_APP_BASE_API_URL}/cards?page=1`
      )
      expect(screen.getByRole('heading')).toHaveTextContent('Rewind')
      expect(screen.getByRole('main')).toMatchSnapshot()
    })

    describe('when searching', () => {
      it('find card name Squirtle', async () => {
        fireEvent.change(screen.getByTestId('searchBox'), {
          target: { value: 'squirtle' },
        })

        mockedAxios.get.mockResolvedValueOnce({
          data: { cards: { data: [cards[2]] }, pagination },
        })

        fireEvent.click(screen.getByText('Search'))

        expect(
          screen.getByText('Loading', { exact: false })
        ).toBeInTheDocument()

        await waitFor(() => screen.getByTestId('searchBox'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/cards?name=squirtle&page=1`
        )

        expect(screen.getByTestId('searchBox')).toHaveDisplayValue('squirtle')

        expect(screen.getByTestId('base493Squirtle0')).toHaveTextContent(
          /squirtle/i
        )
        expect(screen.getAllByRole('row').length).toEqual(2)

        fireEvent.change(screen.getByTestId('searchBox'), {
          target: { value: 'lightning' },
        })

        expect(screen.getByTestId('searchBox')).toHaveDisplayValue('lightning')
      })

      it("reset card's table and URL on blank textbox", async () => {
        fireEvent.click(screen.getByText('Search'))

        await waitFor(() => screen.getByRole('heading'))

        expect(mockedAxios.get).toHaveBeenLastCalledWith(
          `${process.env.REACT_APP_BASE_API_URL}/cards?page=1`
        )
        expect(
          screen.getByTestId('base4128Lightning Energy0')
        ).toHaveTextContent(/lightning/i)
        expect(screen.getByTestId('base493Squirtle0')).toHaveTextContent(
          /squirtle/i
        )
        expect(screen.getAllByRole('row').length).toEqual(4)
      })
    })

    describe('when using contorls', () => {
      describe('to backing up', () => {
        it('click on back up and get the first page back', async () => {
          fireEvent.click(screen.getByTestId('control_backup_create'))

          mockedAxios.post.mockImplementationOnce(async () => {})

          expect(
            screen.getByText('Loading', { exact: false })
          ).toBeInTheDocument()

          await waitFor(() => screen.getByTestId('searchBox'))

          expect(
            screen.getByText('Unable to Retrive Data', { exact: false })
          ).toBeInTheDocument()
        })
      })
    })
  })

  describe('when loading the page', () => {
    it('should not render no card message and display loading text', async () => {
      mockedAxios.get.mockResolvedValueOnce({
        data: { cards: { data: [] }, pagination },
      })

      render(<App />)

      expect(
        screen.queryByText('Loading', { exact: false })
      ).toBeInTheDocument()

      expect(
        screen.queryByText('No Cards Found', { exact: false })
      ).not.toBeInTheDocument()

      await waitFor(() => screen.getByTestId('searchBox'))

      expect(
        screen.queryByText('Loading', { exact: false })
      ).not.toBeInTheDocument()

      expect(
        screen.queryByText('No Cards Found', { exact: false })
      ).toBeInTheDocument()
    })

    describe('on error from the server', () => {
      it('should render an error message', async () => {
        mockedAxios.get.mockRejectedValue(new Error('fake error message'))

        render(<App />)

        await waitFor(() => screen.getByText('error', { exact: false }))

        expect(
          screen.queryByText('No Cards Found', { exact: false })
        ).not.toBeInTheDocument()

        expect(
          screen.getByText('fake error message', { exact: false })
        ).toBeInTheDocument()
      })
    })
  })
})
