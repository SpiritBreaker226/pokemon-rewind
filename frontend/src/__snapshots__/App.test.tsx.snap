// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App fetch card data renders cards component 1`] = `
<main>
  <header
    class="makeStyles-header-3"
  >
    <div
      class="MuiGrid-root MuiGrid-container"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1"
        >
          Pokemon Rewind
        </h1>
      </div>
      <div
        class="MuiGrid-root makeStyles-headerActions-2 MuiGrid-item MuiGrid-grid-xs-4"
      >
        <section>
          <div
            class="MuiFormControl-root MuiTextField-root makeStyles-searchText-4"
          >
            <div
              class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiInput-input"
                data-testid="searchBox"
                id="searchBox"
                name="searchBox"
                placeholder="Search by"
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Search
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </section>
      </div>
      <div
        class="MuiGrid-root makeStyles-headerActions-2 MuiGrid-item MuiGrid-grid-xs-4"
      >
        <section
          class="makeStyles-control-5"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
            data-testid="control_backup_create"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Create Backup
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
            data-testid="control_backup_purge"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiButton-label"
            >
              Purge Backup
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </section>
      </div>
    </div>
  </header>
  <div
    class="makeStyles-appBody-1"
  >
    <section>
      <div
        class="MuiTableContainer-root"
      >
        <table
          class="MuiTable-root"
        >
          <thead
            class="MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                National Pokedex Number
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Subtype
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Supertype
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                HP
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Artist
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Converted Retreat Cost
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Evolves From
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Number
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Rarity
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Series
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Set
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head"
                role="columnheader"
                scope="col"
              >
                Set Code
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base4128Lightning Energy0"
              >
                Lightning Energy
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base4128Energy1"
              >
                Energy
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base4128Basic2"
              >
                Basic
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base41281283"
              >
                128
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base4128Keiji Kinebuchi4"
              >
                Keiji Kinebuchi
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base4128Common5"
              >
                Common
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base4128Base6"
              >
                Base
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base4128Base Set 27"
              >
                Base Set 2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base4128base48"
              >
                base4
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477Jigglypuff0"
              >
                Jigglypuff
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477391"
              >
                39
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477Pokémon2"
              >
                Pokémon
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477Basic3"
              >
                Basic
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477604"
              >
                60
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base47715"
              >
                1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477776"
              >
                77
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477Kagemaru Himeno7"
              >
                Kagemaru Himeno
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477Common8"
              >
                Common
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477Base9"
              >
                Base
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477Base Set 210"
              >
                Base Set 2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base477base411"
              >
                base4
              </td>
            </tr>
            <tr
              class="MuiTableRow-root"
            >
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493Squirtle0"
              >
                Squirtle
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base49371"
              >
                7
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493Pokémon2"
              >
                Pokémon
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493Basic3"
              >
                Basic
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493404"
              >
                40
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base49315"
              >
                1
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493936"
              >
                93
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493Mitsuhiro Arita7"
              >
                Mitsuhiro Arita
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493Common8"
              >
                Common
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493Base9"
              >
                Base
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493Base Set 210"
              >
                Base Set 2
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body"
                data-testid="base493base411"
              >
                base4
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <nav
        aria-label="pagination navigation"
        class="MuiPagination-root makeStyles-pagination-8"
        data-testid="pagination"
      >
        <ul
          class="MuiPagination-ul"
        >
          <li>
            <button
              aria-label="Go to previous page"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded Mui-disabled Mui-disabled"
              disabled=""
              tabindex="-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
                />
              </svg>
            </button>
          </li>
          <li>
            <button
              aria-current="true"
              aria-label="page 1"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded Mui-selected"
              tabindex="0"
              type="button"
            >
              1
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to page 2"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
              tabindex="0"
              type="button"
            >
              2
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to page 3"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
              tabindex="0"
              type="button"
            >
              3
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to page 4"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
              tabindex="0"
              type="button"
            >
              4
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to page 5"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
              tabindex="0"
              type="button"
            >
              5
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <div
              class="MuiPaginationItem-root MuiPaginationItem-ellipsis"
            >
              …
            </div>
          </li>
          <li>
            <button
              aria-label="Go to page 20"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
              tabindex="0"
              type="button"
            >
              20
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
          <li>
            <button
              aria-label="Go to next page"
              class="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page MuiPaginationItem-rounded"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiPaginationItem-icon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </li>
        </ul>
      </nav>
    </section>
  </div>
</main>
`;
